src/common/k8s-api/endpoints/deployment.api.ts

```diff
@@ -5,25 +5,35 @@
 
 import moment from "moment";
 
-import type { IAffinity } from "../workload-kube-object";
-import { WorkloadKubeObject } from "../workload-kube-object";
-import { autoBind } from "../../utils";
+import type { DerivedKubeApiOptions } from "../kube-api";
 import { KubeApi } from "../kube-api";
 import { metricsApi } from "./metrics.api";
-import type { IPodMetrics } from "./pods.api";
-import type { KubeJsonApiData } from "../kube-json-api";
+import type { PodMetricData, PodSpec } from "./pods.api";
 import { isClusterPageContext } from "../../utils/cluster-id-url-parsing";
-import type { LabelSelector } from "../kube-object";
+import type { KubeObjectStatus, LabelSelector } from "../kube-object";
+import { KubeObject } from "../kube-object";
+import { hasTypedProperty, isNumber, isObject } from "../../utils";
 
 export class DeploymentApi extends KubeApi<Deployment> {
+  constructor(opts?: DerivedKubeApiOptions) {
+    super({
+      objectConstructor: Deployment,
+      ...opts ?? {},
+    });
+  }
+
   protected getScaleApiUrl(params: { namespace: string; name: string }) {
     return `${this.getUrl(params)}/scale`;
   }
 
-  getReplicas(params: { namespace: string; name: string }): Promise<number> {
-    return this.request
-      .get(this.getScaleApiUrl(params))
-      .then(({ status }: any) => status?.replicas);
+  async getReplicas(params: { namespace: string; name: string }): Promise<number> {
+    const { status } = await this.request.get(this.getScaleApiUrl(params));
+
+    if (isObject(status) && hasTypedProperty(status, "replicas", isNumber)) {
```

Comment:
Implementation that relies to runtime type checking is bad, super bad.

Reply:
Why? If we get values that at runtime can't be guaranteed to be of a specific type, like here fetching JSON from some url, we _should_ use runtime type checking to verify the type is as expected. (Not saying the implementation here is necessary the best to achieve that).

Reply:
>  If we get values that at runtime can't be guaranteed to be of a specific type, like here fetching JSON from some url

This is not "some url", this is a known api endpoint that returns exactly what we ask. Is runtime checking still good practice? That said, we should handle situations where we get unexpected status code (here it would be something else than 200)... but that's completely different topic and not related to runtime type checks.

Reply:
Sure, there might be use cases for run time type checking but generally said all of those can be achieved with simpler solution. Runtime type checking relies on emitting typing somehow, e.g. via reflect-api. 

Wording on my comment was bad tbh. since as you described, there might be use cases but very limited. Sorry about that.

In here, same could be achieved by `isNumber(status?.replicas)` which is much simpler.

Reply:
@jansav Your solution is not possible because `status` is typed as `unknown` (which is the current best practice within typescript as `any` is considered harmful).

Reply:
All use cases of `any` can be rewritten as `unknown` or `never`. See https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any

Reply:
The usage here is more like adding a typeguard so compiler understands what is the exact type of unknown.

Reply:
imo current approach is better then cast, but depends on the team ofc.

Reply:
Oh I agree, I think that casts should only really be used to downcast.

Reply:
This just tells me that there is even bigger issue with types here which can be solved later.

1. Why is type of `this.request` "any"?
2. Why we are not able to trust that `this.request.get(..)` returns expected type? If it doesn't, it's case for fatal error (should be handled elsewhere). 

But yeah, let's not solve those in this PR, since it requires changes in implementation.

In general, I would limit the influence of `any` to corner (place where it's first time encountered, e.g. API call) and then immediately turn that to type that we expect that place to return. If it doesn't, it's again fatal error. In that way our operational code can work with known types every time and therefore casting or run time type checking is not required.

Reply:
1. `this.request` should not be `any`, where are you seeing this?
2. Because the programmer might have put in the wrong URL, or any other number of bugs. 

What do you mean by limiting `unknown` (I agree, `any` should be banned)? That is exactly what I am doing here, turning it into the type I expect just without a cast (because those are bad in cases like this).

Reply:
1. <img width="1410" alt="image" src="https://user-images.githubusercontent.com/3083083/162671154-cd66da15-2d84-43d6-86dd-b6e5d744f54d.png">

2. Wrong URL would lead to fatal error. As long as URL is not input from user, we should trust it. Why would we add some URL without testing that it works? For real, this is case for automated integration test that verifies that API contract works. 

> What do you mean by limiting unknown (I agree, any should be banned)? That is exactly what I am doing here, turning it into the type I expect just without a cast (because those are bad in cases like this).

Level of abstractions should be correct. At the moment it looks like use places need to prepare for unknown type which isn't correct. There should be function that is called with minimal arguments and returns specific type.